src/helper/Reply.cc

```diff
@@ -151,6 +151,58 @@ isKeyNameChar(char c)
     return false;
 }
 
+/// warns admin about problematic key=value pairs
+void
+Helper::Reply::CheckReceivedKey(const SBuf &key, const SBuf &value)
+{
+    // Squid recognizes these keys (by name) in some helper responses
+    static const std::vector<SBuf> recognized = {
+        SBuf("clt_conn_tag"),
+        SBuf("group"),
+        SBuf("ha1"),
+        SBuf("log"),
+        SBuf("message"),
+        SBuf("nonce"),
+        SBuf("password"),
+        SBuf("rewrite-url"),
+        SBuf("status"),
+        SBuf("store-id"),
+        SBuf("tag"),
+        SBuf("token"),
+        SBuf("url"),
+        SBuf("user")
+    };
+
+    // TODO: Merge with Notes::ReservedKeys(). That list has an entry that Squid
+    // sources do _not_ recognize today ("ttl"), and it is missing some
+    // recognized entries ("clt_conn_tag", "nonce", store-id", and "token").
+
+    if (key.isEmpty()) {
+        debugs(84, DBG_IMPORTANT, "WARNING: Deprecated from-helper annotation without a name: " <<
+               key << '=' << value <<
+               Debug::Extra << "advice: Name or remove this annotation");
+        // TODO: Skip/ignore these annotations.
+        return;
+    }
+
+    // We do not check custom keys for repetitions because Squid supports them:
+    // The "note" ACL checks all of them and %note prints all of them.
+    if (*key.rbegin() == '_')
+        return; // a custom key
+
+    // To simplify, we allow all recognized keys, even though some of them are
+    // only expected from certain helpers or even only in certain reply types.
```

Comment:
> PR description: It is possible to also warn about B2 and B3 annotations, but several draft implementations imply that such an improvement requires complex code changes that may also increase Squid code maintenance overheads.

You can view one such unpolished/unfinished attempt to warn about B2 and B3 annotations at https://github.com/measurement-factory/squid/pull/183. That unofficial PR instruments NotePairs to track explicit Entry "uses" by former findFirst() and find() calls, automatically detecting most B2 and B3 cases. The code is not terrible, and the performance overhead of those changes is relatively small, but the design has significant problems (detailed in the last three paragraphs of that PR description). We also tried two other approaches, but they all had serious shortcomings.

At the end of the day, I think we should start with this simple change and leave further improvements as a long-term goal. If nothing else, this will allow us to collect feedback from admins that use offending helpers. If there is too much negative feedback, we may need to reconsider the current official "just add underscore" approach itself!